Add Climate groups

albertomontesg opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

**Note**: This PR is a continuation of the existing work: https://github.com/home-assistant/core/pull/77737

Adds a climate group component. Based initially on https://github.com/daenny/climate_group, but updated to be similar to the existing light/fan implementations.

Description of the grouping behaviour (taken from the docs PR):

In short, when any group member is active (not `off` state), the most common state is shown.

- The group state is `unavailable` if all group members are `unavailable`.
- Otherwise, the group state is `unknown` if all group members are `unknown` or `unavailable`.
- Otherwise, the group state is `off` if all group members are `off`.
- Otherwise, the group state is the most common state of the group members which are not `off`.

For other modes, the most common mode is reported (e.g. for `PRESET_MODES`, `FAN_MODE` and `SWING_MODE`). The reported temperatures are the averages from the given entities.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #111482
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/32091

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

albertomontesg committed:
Implement Climate group

This would enable grouping different climate entities into a single entity to be controlled together.

This implementation is a fork of the existing Pull request that is right now being unmaintained: https://github.com/home-assistant/core/pull/77737

albertomontesg committed:
Fix original tests to make them pass

rokam commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,342 @@
+        # End temperature settings
+
+        # available HVAC modes
+        all_hvac_modes = list(find_state_attributes(states, ATTR_HVAC_MODES))
```
Should the group only support HVAC that are available for all climate entities?

albertomontesg replied:
Done

rokam commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,342 @@
+            self._attr_hvac_action = None
+
+        # available swing modes
+        all_swing_modes = list(find_state_attributes(states, ATTR_SWING_MODES))
```
Should the group only support swing modes that are available for all climate entities?

albertomontesg replied:
Done

rokam commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,342 @@
+        self._attr_swing_mode = most_frequent_attribute(states, ATTR_SWING_MODE)
+
+        # available fan modes
+        all_fan_modes = list(find_state_attributes(states, ATTR_FAN_MODES))
```
Should the group only support fan mode that are available for all climate entities?

albertomontesg replied:
Done

rokam commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,342 @@
+        self._attr_fan_mode = most_frequent_attribute(states, ATTR_FAN_MODE)
+
+        # available preset modes
+        all_preset_modes = list(find_state_attributes(states, ATTR_PRESET_MODES))
```
Should the group only support presets that are available for all climate entities?

albertomontesg replied:
Done

rokam commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,342 @@
+
+        # Bitwise-and the supported features with the Grouped climate's features
+        # so that we don't break in the future when a new feature is added.
+        self._attr_supported_features &= SUPPORT_FLAGS
```
Should the group only show supported features that are available for all climate entities?

albertomontesg replied:
Done

rokam reviewed (commented):
It looks great. I think that the approach to merge capabilities from all climates should be a little different.

albertomontesg commented:
I think this is a design decision that might be needed to be made (I am not leaning to any specific option here):

1. We define the attributes, their options and capabilities as the **union** of all entities in the group (existing implementation)
2. We define the attributes, their options and capabilities as the **intersection** across all entities in the group (your comments pointing into that direction)

I guess that some of the most common use case might be that people have the same type of devices for their thermostats (what I have at home) and either option would be the same, but in cases where the capabilities between devices are the same, I think option 1. might be better as with option 2, if devices have complementary supported flags might end up creating an entity that does nothing as the intersection of capabilities is empty.

Looking forward on your thoughts @rokam

albertomontesg commented:
Note also that in the same tests with the 2 demo climate devices being grouped, one can set the target temperature and the other can set the temperature range. Merging attributes as a union allows to set both things for each of the different devices while the intersection would make it impossible to set the target temperature and range to either device

rokam commented:
You do get a point. I think that you can follow the same approach as the media_player group. It works as a union but an unsupported action isn't submitted to the entity.

albertomontesg committed:
Update the implementation given feedback provided

Some changes done are:
* Remove the logger as the messages were too simple and other groups do not use it
* Hold the sets of entities that support each set of features, and only make service calls to the entities in the groups that support each method
* Update the logic that defined suppprted features to compute the union of features (rather than hardcode it on the code)
* Made the supported modes in the attributes be the intersection across the entities in the groups (rather than the union)
* Updated tests accordingly

albertomontesg commented:
@rokam originally I misunderstood your comments which I completely agree: the supported modes of the group should be an intersection of the entities, while the supported features should be a union (as of my previous comment).

I have updated the PR with those changes done where the features are the union, but also kept track so we only call the services to the entities that support them, while for the different modes, the intersection is being used.

PTAL

albertomontesg committed:
Merge modes as union and not as intersection

albertomontesg commented:
Checking at other implementations (e.g. [light group](https://github.com/home-assistant/core/blob/045dc3f1fb37a4eb462f1caa5b7e84ab150b9703/homeassistant/components/group/light.py#L273-L278)), it seems to be common that the modes of the group are defined as the union of all its entities, and not intersection, therefore I updated it accordingly.

gjohansson-ST commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,396 @@
+        self._attr_assumed_state |= not states_equal(states)
+
+        # Set group as unavailable if all members are unavailable or missing
+        self._attr_available = any(state.state != STATE_UNAVAILABLE for state in states)
```
If it's unavailable we can probably bail out early?

albertomontesg replied:
Checking at the rest of group implementations there is no early return in case the entity is marked as unavailable. Also conceptually, availability and working attributes of the group entities are independent attributes (if unavailable the entity should not be marked as not having any other attribute).

gjohansson-ST commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,396 @@
+            for x in states
+            if x.state not in [HVACMode.OFF, STATE_UNAVAILABLE, STATE_UNKNOWN]
+        ]
+        # return the most common hvac mode (what the thermostat is set to do) except OFF, UNKNOWN and UNAVAILABE
```
On this one and remaining below. Should we really return the most common one instead of having a sort off priority of states?
Example: One is heating and two are off. Should we then not return `heat` instead of `off`?

albertomontesg replied:
HVAC modes used as state of a climate entity can be seen as an binary state with multiple ON states. Here the logic is filtering out the negative stats (OFF, UNAVAILABLE) so if any entity state is active (or ON state) those values are taken into account. Then similarly on how some enums attributes are being aggregated across entities in other groups, taking the most common enum value has been the consensus at the moment.

gjohansson-ST replied:
Ok. So perhaps I took a bad example so let's do another one.
3 entities, one is `heat` and two are `fan_only`. With your logic the group state would be `fan_only`.
Is that really what is wished here instead of returning `heat`?

Btw: Don't resolve until we agree, it's annoying to have to unresolve to continue the discussion

albertomontesg replied:
> 3 entities, one is heat and two are fan_only. With your logic the group state would be fan_only.
Is that really what is wished here instead of returning heat?

Yes that was my intention since the beginning. I think the logic behind of why `heat` might be more important than `fan_only` could be driven that could be a more costly mode. This issue was raised on the discussion of the original  docs PR: https://github.com/home-assistant/home-assistant.io/pull/23990#discussion_r973765098

That being said, I believe that most common mode might be the fairest approach as ranking can be really hard to do. `heat` might be ranked higher than `cool` or `fan_only` on colder countries, while it could be the other way around on much more hot countries. This can be a matter of perspective depending on how costly each mode is on different climate environments, therefore having a more uniform approach looks the right solution to me by taking the most common mode across all working modes (excluding `off`).

gjohansson-ST commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,396 @@
+            self._attr_hvac_action = None
+
+        # available swing modes
+        all_swing_modes = list(find_state_attributes(states, ATTR_SWING_MODES))
+        if all_swing_modes:
+            self._attr_swing_modes = merge_modes(all_swing_modes)
```
```suggestion
        self._attr_swing_modes = list(set(find_state_attributes(states, ATTR_SWING_MODES)))
```
I suppose something like this should do? (and the `merge_modes` function is not needed?)

albertomontesg replied:
Note that `find_state_attributes` would return a list (one entry for each state) of lists (with all the modes). The suggested changed would yield a list of lists (note that lists are hashable and can be used inside a set). For that reason the logic `list(set().union(*modes))` is used. Wrapped it in a simple method for readability purposes given that it is also being used multiple times.

gjohansson-ST replied:
Right, my point was more that this needs to be able to be set as an empty list.
Now if nothing comes back it will just skip it and `swing_modes` will be the previous list.

albertomontesg replied:
Done

gjohansson-ST commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,396 @@
+                supported_features |= feature_flags
+        self._attr_supported_features = supported_features
+
+    async def async_set_temperature(self, **kwargs: Any) -> None:
```
I wonder a bit about this one as we can send temp-range to entities which doesn't support it.
Should we not be more explicit about only sending what is actually supported?

albertomontesg replied:
Given that we are at the end calling the same method/service for all entities with all data (target value and range), each entity implementation would consume the values it needs ([example](https://github.com/home-assistant/core/blob/dev/homeassistant/components/ecobee/climate.py#L703-L707)), we can have here a simple logic that would call the same service with all the available data to all the entities in the group

gjohansson-ST replied:
I don't think you entirely understand my point.
Looking at `sensibo` as example it doesn't support `target_range` and therefore it would raise if you call it with range values, see https://github.com/home-assistant/core/blob/dev/homeassistant/components/sensibo/climate.py#L321

I don't think that's very user friendly

albertomontesg replied:
Agreed that would not be very user friendly, but I would argue that here we are talking about a more fundamental API problem. Having a method that contains a freeform of arguments being passed (through the `**kwargs`) without any guarantees about which parameters are being passed independently of the supported features of each entity type, leads to individual implementations handling their optionality in different ways.

With some [search](https://github.com/search?q=repo%3Ahome-assistant%2Fcore+symbol%3Aasync_set_temperature+path%3Aclimate&type=code&p=2), you can see that there are different implementations handling the presence and missing of parameters in different ways:

1. If not getting the value you need, do nothing: [code](https://github.com/home-assistant/core/blob/b0a1450a2b654d338d6fee7033b675b73275cb1a/homeassistant/components/generic_thermostat/climate.py#L395)
2. Silently raising a `KeyError` exception: [examples](https://github.com/search?q=repo%3Ahome-assistant%2Fcore+symbol%3Aasync_set_temperature+path%3Aclimate+%22kwargs%5BATTR_TEMPERATURE%5D%22&type=code)
3. Pass the target temperature value or `None` down the line to the underlying coordinator: [code](https://github.com/home-assistant/core/blob/f0b07ae942317aea31e24741d4c211ac065d3eaf/homeassistant/components/atag/climate.py#L86)
4. And others around how to handle the temperature range I did not went too far to dig in....

The point you just raise about the exception being raised would have also happened if we were to make a service call to a list of entity types and providing all parameters in the schema (at the end a group is doing that when its service is being called).  If we call a `sensibo` device with an `ecobee` with all parameters in the schema, we would get the methods called, but because [of this line](https://github.com/home-assistant/core/blob/f01235ef746823a8f146c603bf732df6a6061706/homeassistant/helpers/service.py#L996) we would not be calling the methods to update the state in the state machine, which again is not very user friendly.

My point (and preference) here is that in an ideal world, logic should be simple everywhere rather than complex. Specific climate platforms should not raise Exceptions if not supported parameters are being passed, as well as the group should just forward whatever parameters have been given to all the entities in the group (more future proof in case the schema of the call gets extended).

I can update the logic and make it more complex to take into account the supported features for each entity in the group so we make a targeted call for each entity, but this would not solve the bigger underlying problem that set temperature in a Climate entity ID already has and this change is not enlarging.

Let me know if I should proceed with updating the logic.

albertomontesg replied:
Alternatively, the logic to decide which parameters should be passed to the `async_temperature_set` method given the supported features, could be placed [here](https://github.com/home-assistant/core/blob/f01235ef746823a8f146c603bf732df6a6061706/homeassistant/components/climate/__init__.py#L902-L917) which would take affect across the board for all Climate entities to do the right thing and avoid the mentioned problem. This most likely should be a separate PR.

albertomontesg replied:
Proposed https://github.com/home-assistant/core/pull/114538 to handle this logic more natively on the ClimateEntity implementation.

gjohansson-ST replied:
the schema handles that we can't use temperature and a temp range at the same time.
So my thinking was simply to check if it's a single temp or a temp range, then check the corresponding supported entities and send the call only to the entities which has support.

note that all these are optional from the schema so you need to something like
```python
if kwargs.get(ATTR_TEMPERATURE):
    data = {
            ATTR_ENTITY_ID: self._features[
                ClimateEntityFeature.TARGET_TEMPERATURE
            ]
        }
elif kwargs.get(ATTR_TARGET_TEMP_LOW):
    data = {
            ATTR_ENTITY_ID: self._features[
                ClimateEntityFeature.TARGET_TEMPERATURE_RANGE
            ]
        }
...
```

albertomontesg replied:
Can we bring the discussion about what arguments to pass to the `set_temperature` method to the architectural proposal (https://github.com/home-assistant/architecture/discussions/1071) given that this logic seems more appropriate to be defined at the base ClimateEntity logic rather than on this children specialisation. The proposal would remove any need to handle this logic on this change.

gjohansson-ST commented on homeassistant/components/group/config_flow.py:
```diff
@@ -183,6 +184,11 @@ async def _set_group_type(
     ),
+    "climate": SchemaFlowFormStep(
+        basic_group_config_schema("climate"),
+        preview="climate",
```
I didn't see any preview method in the climate entity

albertomontesg replied:
Added the preview and also test it

gjohansson-ST reviewed (changes requested):
Some initial comments.

I think the logic needs to be described much better in the PR description as well.
I see you had some feedback in the docs PR.

albertomontesg commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,396 @@
+        self._name = name
+        self._entity_ids = entity_ids
+
+        self._features: dict[ClimateEntityFeature, set[str]] = {
```
Because as other group implementations, this dictionary is holding the entities in the groups that support each features or service, and then use to just call the service on top of the entities that support the feature (rather than call the service on top of all entities in the group).

albertomontesg commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,396 @@
+        vol.Required(CONF_ENTITIES): cv.entities_domain(DOMAIN),
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_UNIQUE_ID): cv.string,
+        vol.Optional(CONF_TEMPERATURE_UNIT): cv.temperature_unit,
```
Done

albertomontesg committed:
Added preview for the config flow for the climate group

albertomontesg commented:
Added the behaviour of the group into the PR description, as well as submitted a new PR for the docs update (original change is from 1.5y ago). Also addressed all the comments from the review.

albertomontesg committed:
Updates attributes behavior given feedback

albertomontesg committed:
Selectively send the correct temperature arguments to the different subentities

gjohansson-ST commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,390 @@
+            # we find.
+            self._attr_supported_features |= support
+
+        # Bitwise-and the supported features with the Grouped climate's features
+        # so that we don't break in the future when a new feature is added.
+        all_supported_features = ClimateEntityFeature(0)
+        for feature_flags in self._features:
+            all_supported_features |= feature_flags
+        self._attr_supported_features &= all_supported_features
```
Why not adding new features if available by input entities?

albertomontesg replied:
The set of features present in the `_features` dictionary are the ones that this implementation is supporting (by implementing corresponding methods from ClimateEntity). We don't want to report the group to support a feature (that could be added in the future) without explicitly updating this group implementation to support it. This logic therefore restricts that the supported features are the intersection between the ones from the input entities and the ones the group class has implemented.

gjohansson-ST replied:
Why not adding it to `self._features` instead to allow dynamically expanding with new features.
That was what I meant with one of my very first comments on `self._features` that why isn't it constructed dynamically for everything possible from `ClimateEntityFeature`

albertomontesg replied:
Because with the definition of the keys in a dynamic way, we can hit in the future that a new feature is added, and when a feature is added a method also needs to be implemented with its corresponding logic. This can happen that in the future some one else adds a new feature, this gets dynamically added as a key to the `_features` dict, and then when called the service to the ClimateGroup that is related to that feature, we would get some error about not being implemented. Therefore is safer to be explicit about which are the features that the `ClimateGroup` is supporting rather than expect it to be dynamically defined without any changes on the logic.

albertomontesg replied:
It was not setting the variable twice but computing the intersection between the features from the entities and the ones supported by the ClimateGroup class.

Indeed the logic could be reduced to fewer lines of code so I did this.

gjohansson-ST commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,390 @@
+
+    async def async_set_temperature(self, **kwargs: Any) -> None:
+        """Forward the temperature command to all climate in the climate group."""
+        if kwargs.get(ATTR_TEMPERATURE):
+            data = {
+                ATTR_ENTITY_ID: self._features[ClimateEntityFeature.TARGET_TEMPERATURE],
+                ATTR_TEMPERATURE: kwargs[ATTR_TEMPERATURE],
+            }
```
```suggestion
        if (temp := kwargs.get(ATTR_TEMPERATURE)):
            data = {
                ATTR_ENTITY_ID: self._features[ClimateEntityFeature.TARGET_TEMPERATURE],
                ATTR_TEMPERATURE: temp,
            }
```

albertomontesg replied:
Done

gjohansson-ST commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,390 @@
+                ATTR_ENTITY_ID: self._features[ClimateEntityFeature.TARGET_TEMPERATURE],
+                ATTR_TEMPERATURE: kwargs[ATTR_TEMPERATURE],
+            }
+        elif kwargs.get(ATTR_TARGET_TEMP_HIGH):
+            data = {
+                ATTR_ENTITY_ID: self._features[ClimateEntityFeature.TARGET_TEMPERATURE],
+                ATTR_TARGET_TEMP_HIGH: kwargs[ATTR_TARGET_TEMP_HIGH],
+                ATTR_TARGET_TEMP_LOW: kwargs[ATTR_TARGET_TEMP_LOW],
+            }
```
```suggestion
        elif (temp_high := kwargs.get(ATTR_TARGET_TEMP_HIGH)) and (temp_low := kwargs.get(ATTR_TARGET_TEMP_LOW)):
            data = {
                ATTR_ENTITY_ID: self._features[ClimateEntityFeature.TARGET_TEMPERATURE],
                ATTR_TARGET_TEMP_HIGH: temp_high,
                ATTR_TARGET_TEMP_LOW: temp_low,
            }
```

albertomontesg replied:
Done

gjohansson-ST commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,390 @@
+                ATTR_TARGET_TEMP_HIGH: kwargs[ATTR_TARGET_TEMP_HIGH],
+                ATTR_TARGET_TEMP_LOW: kwargs[ATTR_TARGET_TEMP_LOW],
+            }
+        if ATTR_HVAC_MODE in kwargs:
+            data |= {
+                ATTR_HVAC_MODE: kwargs[ATTR_HVAC_MODE],
+            }
```
```suggestion
        if (hvac_mode := kwargs.get(ATTR_HVAC_MODE)):
            data |= {
                ATTR_HVAC_MODE: hvac_mode,
            }
```

albertomontesg replied:
Done

albertomontesg committed:
Style improvements

albertomontesg committed:
Reduce the logic when computing supported features

albertomontesg commented:
@gjohansson-ST any further blocker for the PR to be approved?

r0bb10 commented:
this is great, the original custom componet work seems to be broken with 2024.4 so i'll wait for this to be merged!

albertomontesg commented:
@gjohansson-ST @rokam friendly ping to finalize the review for this PR.

gjohansson-ST commented:
Please stop pinging for reviews. I'll come back to this in due time.
Also there is lacking coverage so please make sure CI passes.

albertomontesg committed:
Extend tests to cover all services exposed and their logic

In addition handled the propagation of temperature for all devices either if they support target temperature or temperature range.

albertomontesg commented:
Extended the tests to cover all service calls for this Group entity implementation.

albertomontesg committed:
Add Missing string on Menu options

gjohansson-ST commented:
Just wanted to provide some feedback here. Probably the PR is fine and I will do some testing with it to make sure it works as intended. 
But we have another open discussion on group states which I think we need to sort first before bringing in this one (as it's about how we handle multiple on-states). So just you are aware that this one unfortunately needs to be delayed a bit until that one is finally decided and implemented.

albertomontesg commented:
The current implementation handles the on states the same way as other existing group components. Is it strictly necessary to merge this after that discussion is sorted out, or could it be merge before and then implement whatever changes are needed? This way we could have the functionality available to users possibly some releases before there is a solution for the groups states.

Is there any discussion about the on states for groups that I can follow?

albertomontesg commented:
@gjohansson-ST seeing the number of reactions on the previous comment, can the decision to delay this PR be reconsidered? Updating the logic about whether how multiple on states should be handled does not only relate to Climate but other existing groups, therefore seems orthogonal to the goal of this change which is adding a missing functionality into Home Assistant.

marcelhoogantink commented:
Is this still ilve??

alexbbt commented:
I want to echo @albertomontesg's thoughts from above. I have been following this PR for the last few months as it solves from pretty big issues for me. I have multiple automations and scripts setup to create a "master" thermostat for my 4 minisplits and would really love to just use a group. I understand that groups might be changing, and it actually sounds great, but in the mean time, it would be so helpful to get this update merged. Then whoever is working on the new group logic, can consider climate groups at the same time.

jafar-atili commented:
@albertomontesg, can you please create a custom integration for the meanwhile?

frenck commented:
HI there @albertomontesg 👋 

We've discussed this PR with our core team. This PR, similar PRs in the past, and feature requests have been around for quite some time now; showing the definitely need for it.

From a generic point of view, we agree on adding this feature to Home Assistant. So, from that perspective, it is a go. But we do have concerns preventing us from adding this feature.

This PR is questionable on some points for us. Therefore, we'd like to ask you to write up how a group would behave. The PR description contains the basics for the main entity state but not how everything else would behave.

For example, let's say we combine thermostats that can only cool, only heat with one that supports heat/cool (and thus a range). Now, what would that mean for temperature control, HVAC control and preset control?

Note, this is just one example, but we do need to clearly decide (and eventually also clearly document) the expected behavior of this (and validate if this actually is the expected behavior from an end-user perspective).

../Frenck

Wazbat commented:
> For example, let's say we combine thermostats that can only cool, only heat with one that supports heat/cool (and thus a range). Now, what would that mean for temperature control, HVAC control and preset control?

Maybe only shared options should be available?
For example if there are thermostats that can only heat and cool, and another that can only heat, then the only available options are to heat?

Alternatively, maybe all options should be available, with configurable presets? For example if your radiators can only heat, and the AC unit can cool, you can configure the radiators to turn fallback to `off` when an unsupported mode is set for example?

frenck commented:
> Maybe only shared options should be available?

That can't work, as these can change during the lifecycle.

marcelhoogantink commented:
It seems that @albertomontesg has lost motivation, due to the lack of structural positive replies by the reviewing dev's.

It is something I also experienced 😟

frenck commented:
While awaiting a response to the above; I've drafted the PR for now.

The reason being, that is shows up as "ready for review" in our queues, while we are actually awaiting a response. Thus, marking it draft for a bit.

bcutter commented:
@albertomontesg thank you for taking care of this.

With HA Core 2024.1 update, the former/current custom integration starts to create quite some warnings (no issue exists for this as the custom integration is unmaintained as statet at https://github.com/daenny/climate_group?tab=readme-ov-file#climate_group).

```
Logger: homeassistant.components.climate.const
Source: helpers/deprecation.py:205
Integration: Klima (documentation, issues)
First occurred: 16:15:21 (20 occurrences)
Last logged: 16:15:21

SUPPORT_TARGET_HUMIDITY was used from climate_group, this is a deprecated constant which will be removed in HA Core 2025.1. Use ClimateEntityFeature.TARGET_HUMIDITY instead, please report it to the author of the 'climate_group' custom integration
SUPPORT_FAN_MODE was used from climate_group, this is a deprecated constant which will be removed in HA Core 2025.1. Use ClimateEntityFeature.FAN_MODE instead, please report it to the author of the 'climate_group' custom integration
SUPPORT_PRESET_MODE was used from climate_group, this is a deprecated constant which will be removed in HA Core 2025.1. Use ClimateEntityFeature.PRESET_MODE instead, please report it to the author of the 'climate_group' custom integration
SUPPORT_SWING_MODE was used from climate_group, this is a deprecated constant which will be removed in HA Core 2025.1. Use ClimateEntityFeature.SWING_MODE instead, please report it to the author of the 'climate_group' custom integration
SUPPORT_AUX_HEAT was used from climate_group, this is a deprecated constant which will be removed in HA Core 2025.1. Use ClimateEntityFeature.AUX_HEAT instead, please report it to the author of the 'climate_group' custom integration
```

Therefore (clock is ticking 😃) I appreciate your work. Thanks for following up on this (as requested by frenck at https://github.com/home-assistant/core/pull/114290#issuecomment-2180386218).

emontnemery commented:
@albertomontesg if you want the PR to be reviewed, you need to click the "ready for review" button

frenck commented:
Alright, we had a bunch of questions open, which, unfortunately, have been unanswered at this point. While we are open to this feature, this one isn't ready at this point.

Closing this PR for being stale right now. Feel free to re-open when you are ready to pick up work on it again.

../Frenck
